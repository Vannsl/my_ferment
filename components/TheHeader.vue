<template>
  <nav :class="{ 'bg-white shadow': isSticky }" class="fixed w-full z-30 top-0">
    <div class="w-full container mx-auto flex flex-wrap items-center justify-between mt-0 py-2">
      <div class="pl-4 flex items-center">
        <logo :isStickable="true" :isSticky="isSticky" />
      </div>
      <div class="block lg:hidden pr-4">
        <button
          :class="{ 'bg-primary': isSticky }"
          class="flex items-center px-3 py-2 border rounded text-white border-white hover:text-gray-800 hover:border-gray-800 appearance-none focus:outline-none"
          @click.prevent.stop="onToggleClick"
        >
          <hamburger-svg />
        </button>
      </div>

      <div
        :class="[isSticky ? 'bg-white' : 'bg-gray-100', isOpen ? 'hidden' : '']"
        class="w-full flex-grow lg:flex lg:items-center lg:w-auto lg:block mt-2 lg:mt-0 bg-white lg:bg-transparent text-black p-4 lg:p-0 z-20"
      >
        <ul class="list-reset lg:flex justify-end flex-1 items-center">
          <!-- <li class="mr-3">
            <a href="#howto">Test</a>
           <nuxt-link
              class="link hover:text-gray-800 hover:text-underline"
              :to="{ path: '/', hash:'howto'}"
            >
              {{ $t('navigation.new') }}
            </nuxt-link>
          </li>
          <li class="mr-3">
            <nuxt-link
              class="link hover:text-gray-800 hover:text-underline"
              to="/"
            >{{ $t('navigation.recipes') }}</nuxt-link>
          </li> -->
          <li class="mr-3">
            <a
              class="link hover:text-gray-800 hover:text-underline"
              href="#"
            >{{ $t('navigation.seasonalCalendar') }}</a>
          </li>
        </ul>
        <nuxt-link
          :class="[ isSticky ? 'bg-primary text-white' : 'bg-white text-gray-800' ]"
          class="mx-auto lg:mx-0 hover:underline font-bold rounded-full mt-4 lg:mt-0 py-4 px-8 shadow opacity-75"
          to="/login"
        >{{ $t('navigation.myFerments') }}</nuxt-link>
      </div>
    </div>
    <hr class="border-b border-gray-100 opacity-25 my-0 py-0" />
  </nav>
</template>

<script>
import { mapState } from 'vuex'
import Logo from '@/components/Logo'
import HamburgerSvg from '@/components/svg/Hamburger'

export default {
  name: 'TheHeader',
  components: {
    logo: Logo,
    'hamburger-svg': HamburgerSvg
  },
  data() {
    return {
      scrollY: 0,
      isOpen: false
    }
  },
  computed: {
    ...mapState('auth', ['authenticatedUser']),
    isSticky() {
      return this.scrollY > 10
    }
  },
  methods: {
    onClick() {
      this.isOpen = false
    },
    onScroll() {
      this.scrollY = window.scrollY
    },
    onToggleClick() {
      this.isOpen = !this.isOpen
    }
  },
  mounted() {
    this.scrollY = window.scrollY
    document.addEventListener('click', this.onClick)
    document.addEventListener('scroll', this.onScroll)
  },
  beforeDestroy() {
    document.removeEventListener('click', this.onClick, true)
    document.removeEventListener('scroll', this.onScroll, true)
  }
}
</script>

<style scoped>
.link {
  @apply inline-block text-black no-underline py-2 px-4;
}
</style>
