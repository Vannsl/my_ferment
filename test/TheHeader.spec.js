import { shallowMount } from '@vue/test-utils'
import TheHeader from '@/components/TheHeader'

const clickEvent = () => {
  const event = new MouseEvent('click', {
    view: window,
    bubbles: true,
    cancelable: true
  })
  document.dispatchEvent(event)
}

const scrollEvent = () => {
  const event = new CustomEvent('scroll')
  document.dispatchEvent(event)
}

const onClickSpy = jest.fn()
const onScrollSpy = jest.fn()

const factory = () => {
  return shallowMount(TheHeader)
}

const spiedFactory = () => {
  return shallowMount(TheHeader, {
    attachToDocument: true,
    methods: { onClick: onClickSpy, onScroll: onScrollSpy }
  })
}

describe('TheHeader', () => {
  it('is a Vue instance', () => {
    const wrapper = factory()
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('renders properly', () => {
    const wrapper = factory()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('sets the attribute scrollY to 0', () => {
    const wrapper = factory()
    expect(wrapper.vm.scrollY).toBe(0)
  })

  it('sets the attribute isOpen to false', () => {
    const wrapper = factory()
    expect(wrapper.vm.isOpen).toBe(false)
  })

  describe('when mounted', () => {
    let addEventSpy

    beforeEach(() => {
      addEventSpy = spyOn(document, 'addEventListener')
    })

    it('adds a click event listener', () => {
      const wrapper = factory()
      expect(addEventSpy).toHaveBeenCalledWith('scroll', wrapper.vm.onScroll)
    })

    it('adds a scroll event listener', () => {
      const wrapper = factory()
      expect(addEventSpy).toHaveBeenCalledWith('click', wrapper.vm.onClick)
    })
  })

  describe('when destroyed', () => {
    let wrapper
    let removeEventSpy

    beforeEach(() => {
      wrapper = factory()
      removeEventSpy = spyOn(document, 'removeEventListener')
    })

    it('removes a click event listener', () => {
      wrapper.destroy()
      expect(removeEventSpy).toHaveBeenCalledWith(
        'scroll',
        wrapper.vm.onScroll,
        true
      )
    })

    it('removes a scroll event listener', () => {
      wrapper.destroy()
      expect(removeEventSpy).toHaveBeenCalledWith(
        'click',
        wrapper.vm.onClick,
        true
      )
    })
  })

  describe('when the user clicks on the hamburger menu', () => {
    it('onToggleClick is triggered', () => {
      const wrapper = factory()
      const spy = spyOn(wrapper.vm, 'onToggleClick')
      wrapper.find('button').trigger('click')
      expect(wrapper.vm.onToggleClick).toBeCalled()
    })

    it('sets the property isOpen to true', () => {
      const wrapper = factory()
      expect(wrapper.vm.isOpen).toBe(false)
      wrapper.vm.onToggleClick()
      expect(wrapper.vm.isOpen).toBe(true)
    })

    it('displays the actions', done => {
      const wrapper = factory()
      expect(wrapper.vm.navContentClassList).toEqual(
        expect.stringContaining('hidden')
      )
      wrapper.find('button').trigger('click')
      wrapper.vm.$nextTick(() => {
        expect(wrapper.vm.navContentClassList).toEqual(
          expect.not.stringContaining('hidden')
        )
        done()
      })
    })
  })

  describe('when the user clicks outside of the burger menu', () => {
    it('onClick is triggered', () => {
      const wrapper = spiedFactory()
      clickEvent()
      expect(onClickSpy).toBeCalled()
    })

    it('sets the property isOpen to false', () => {
      const wrapper = factory()
      wrapper.vm.onClick()
      expect(wrapper.vm.isOpen).toBe(false)
    })

    it('hides the actions', done => {
      const wrapper = factory()
      clickEvent()
      wrapper.vm.$nextTick(() => {
        expect(wrapper.vm.navContentClassList).toEqual(
          expect.stringContaining('hidden')
        )
        done()
      })
    })
  })

  describe('when the user clicks scrolls on the page', () => {
    it('onScroll is triggered', () => {
      const wrapper = spiedFactory()
      scrollEvent()
      expect(onScrollSpy).toBeCalled()
    })

    it('sets the property isOpen to false', () => {
      const wrapper = factory()
      window.scrollY = 10
      scrollEvent()
      expect(wrapper.vm.scrollY).toBe(10)
    })

    describe('<= 10px scrollY', () => {
      it('does not add the sticky classes to the header', () => {
        const wrapper = factory()
        wrapper.setData({ scrollY: 10 })
        expect(wrapper.vm.headerClassList).toEqual(
          expect.not.stringContaining('bg-white shadow')
        )
      })

      it('does not add the sticky classes to the navActions', () => {
        const wrapper = factory()
        wrapper.setData({ scrollY: 10 })
        expect(wrapper.vm.navActionClassList).toEqual(
          expect.not.stringContaining('gradient text-white')
        )
        expect(wrapper.vm.navActionClassList).toEqual(
          expect.stringContaining('bg-white text-gray-800')
        )
      })

      it('does not add the sticky classes to the navContent', () => {
        const wrapper = factory()
        wrapper.setData({ scrollY: 10 })
        expect(wrapper.vm.navContentClassList).toEqual(
          expect.not.stringContaining('bg-white')
        )
        expect(wrapper.vm.navContentClassList).toEqual(
          expect.stringContaining('bg-gray-100')
        )
      })
    })

    describe('> 10px scrollY', () => {
      it('adds the sticky classes to the header', () => {
        const wrapper = factory()
        wrapper.setData({ scrollY: 11 })
        expect(wrapper.vm.headerClassList).toEqual(
          expect.stringContaining('bg-white shadow')
        )
      })

      it('adds the sticky classes to the navActions', () => {
        const wrapper = factory()
        wrapper.setData({ scrollY: 11 })
        expect(wrapper.vm.navActionClassList).toEqual(
          expect.stringContaining('gradient text-white')
        )
        expect(wrapper.vm.navActionClassList).toEqual(
          expect.not.stringContaining('bg-white text-gray-800')
        )
      })

      it('adds the sticky classes to the navContent', () => {
        const wrapper = factory()
        wrapper.setData({ scrollY: 11 })
        expect(wrapper.vm.navContentClassList).toEqual(
          expect.stringContaining('bg-white')
        )
        expect(wrapper.vm.navContentClassList).toEqual(
          expect.not.stringContaining('bg-gray-100')
        )
      })
    })
  })
})
