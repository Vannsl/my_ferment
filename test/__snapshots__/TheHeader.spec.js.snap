// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheHeader renders properly 1`] = `"<nav class=\\"fixed w-full z-30 top-0\\"><div class=\\"w-full container mx-auto flex flex-wrap items-center justify-between mt-0 py-2\\"><div class=\\"pl-4 flex items-center\\"><logo-stub isstickable=\\"true\\"></logo-stub></div> <div class=\\"block lg:hidden pr-4\\"><button class=\\"flex items-center px-3 py-2 border rounded text-gray-500 border-gray-600 hover:text-gray-800 hover:border-teal-500 appearance-none focus:outline-none\\"><hamburger-svg-stub></hamburger-svg-stub></button></div> <div class=\\"w-full flex-grow lg:flex lg:items-center lg:w-auto lg:block mt-2 lg:mt-0 bg-white lg:bg-transparent text-black p-4 lg:p-0 z-20 bg-gray-100 hidden\\"><ul class=\\"list-reset lg:flex justify-end flex-1 items-center\\"><li class=\\"mr-3\\"><a href=\\"#\\" class=\\"inline-block py-2 px-4 text-black font-bold no-underline\\">Active</a></li> <li class=\\"mr-3\\"><a href=\\"#\\" class=\\"inline-block text-black no-underline hover:text-gray-800 hover:text-underline py-2 px-4\\">link</a></li> <li class=\\"mr-3\\"><a href=\\"#\\" class=\\"inline-block text-black no-underline hover:text-gray-800 hover:text-underline py-2 px-4\\">link</a></li></ul> <button class=\\"mx-auto lg:mx-0 hover:underline font-bold rounded-full mt-4 lg:mt-0 py-4 px-8 shadow opacity-75 bg-white text-gray-800\\">Action</button></div></div> <hr class=\\"border-b border-gray-100 opacity-25 my-0 py-0\\"></nav>"`;
